pkg/ccl/importccl/read_import_mysqlout_test.go

```diff
@@ -0,0 +1,312 @@
+		{i: 20, s: `a quote " or two quotes "" and a quote-comma ", , and then a quote and newline "` + "\n"},
+		{i: 21, s: `"a slash \, a double slash \\, a slash+quote \",  \` + "\n"},
+	}
+
+	for i := 0; i < 10; i++ {
+		buf := make([]byte, 200)
+		r.Seed(int64(i))
+		r.Read(buf)
+		testRows = append(testRows, testRow{i: i + 100, s: randStr(r, badChars, 1000), b: buf})
+	}
+
+	configs := []dumpCfg{
+		{
+			name: "escape-and-enclose",
+			opts: roachpb.MySQLOutfileOptions{
+				FieldSeparator: '\t', RowSeparator: '\n', HasEscape: true, Escape: '\\', Enclose: roachpb.MySQLOutfileOptions_Always, Encloser: '"',
+			},
+			null: `\N`,
+		},
+		{
+			name: "csv-ish",
+			opts: roachpb.MySQLOutfileOptions{
+				FieldSeparator: ',', RowSeparator: '\n', HasEscape: true, Escape: '\\', Enclose: roachpb.MySQLOutfileOptions_Always, Encloser: '"',
+			},
+			null: `\N`,
+		},
+		{
+			name: "escape-quote-csv-no-enclose",
+			opts: roachpb.MySQLOutfileOptions{FieldSeparator: ',', RowSeparator: '\n', HasEscape: true, Escape: '"'},
+			null: `\N`,
+		},
+	}
+
+	for i := range configs {
+		configs[i].filename = filepath.Join(`testdata`, `mysqlout`, configs[i].name, `test.txt`)
+	}
+	// This can be toggled to re-write the `testdata`. Requires local mysql
+	// configured with no password access for `root`, a db called `test` and
+	// `OUTFILE` enabled. On OSX, this can be setup using `brew install mysql` and
+	// the following ~/.my.cnf:
+	// [mysqld_safe]
+	// [mysqld]
+	// secure_file_priv=""
+	if false {
+		writeMysqlOutfileTestdata(t, testRows, configs)
+	}
+	return testRows, configs
+}
+
+func TestMysqlOutfileReader(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	testRows, configs := getMysqlOutfileTestdata(t)
+	ctx := context.TODO()
+	for _, config := range configs {
+		t.Run(config.name, func(t *testing.T) {
+			converter := newMysqloutfileReader(ctx, nil, config.opts, nil, 3)
+			// unblock batch chan sends
+			converter.csvInputReader.recordCh = make(chan csvRecord, 4)
+			converter.csvInputReader.batchSize = 10
+
+			in, err := os.Open(config.filename)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer in.Close()
+			if err := converter.readFile(ctx, in, 1, config.name, func(_ bool) error { return nil }); err != nil {
+				t.Fatal(err)
+			}
+			converter.inputFinished()
+
+			csvVersion := filepath.Join(filepath.Dir(config.filename), `test.out.csv`)
+			csvOut, err := os.Create(csvVersion)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer csvOut.Close()
+			w := csv.NewWriter(csvOut)
+
+			var res [][]string
+			var batches int
+			for b := range converter.csvInputReader.recordCh {
+				batches++
+				for _, r := range b.r {
+					if len(r) != 3 {
+						t.Fatalf("bad row len for %v", r)
+					}
+					// If the csv writing part of our conversion encodes as hex, we can
+					// round-trip bytes though the go csv parser.
+					orig := r[2]
+					r[2] = hex.EncodeToString([]byte(r[2]))
```

Comment:
i don't understand what's going on here. why don't we have to append `\x` to this?